### DISCUSSION TITLE
Dynamically add parameters from part creation window?

### DISCUSSION THREAD

**Discussion by isaacolsen94** [inventree/InvenTree#6708] [Q&A]
Hey, new here! I've been playing around with a few of the different inventory self hosted options. When I was playing with Part-DB it was pretty straight forward to search digikey, add the part, download the datasheet, create the parameters from the digikey import and add more if wanted. Then move and create a storage location from the part. The only thing I didn't like was that I couldn't bulk modify.

Long story short I just spun up a docker for Inventree and I'm wondering if all the above is possible?

I see there is a Digikey plug in, but couldn't find a location in the GUI to do anything with it? I also couldn't find a option to add a new storage location or new parameter from the part? Last I couldn't find any information on uploading attachments, but saw that it says it's available.

Any help would be appreciated!

**Comment by SchrodingersGat** [comment]
@isaacolsen94 welcome aboard!

Adding parameters to a new part must be performed *after* the part is created. This can be done via the web interface, or the API, and as such you can use external tools such as [ki-n-tree](https://github.com/sparkmicro/Ki-nTree) to perform this data entry.

**Reply by isaacolsen94** [reply]
Hey thanks! Glad to be here! 

Oh okay that makes sense why I was having issues! :P And that tool looks great. I'll give that a shot. 

Does the same rule with parameters apply to uploads? So I have to save the part before I can upload the attachments?

**Reply by isaacolsen94** [reply]
Also is there a way to add a bulk amount of parts? 

Ex: I have 100 resistor values that are all the same except for different resistance/quantity. Can I create them all at once or do I need to create each one individually?

**Reply by SchrodingersGat** [reply]
You could use an external script (via the [python interface](https://docs.inventree.org/en/stable/api/python/python/) to automate this for you.

Or, there may be a plugin (such as https://inventree.org/wolflu05/inventree-bulk-plugin) which would do this? 

And you can also [import data](https://docs.inventree.org/en/stable/settings/import/) from an external file such as a spreadsheet or CSV

**Reply by wolflu05** [reply]
Yes, you can use the [bulk plugin](https://github.com/wolflu05/inventree-bulk-plugin) like referenced by oliver to bulk generate parts as well. This gets pretty handy for such electrical components that exist in different values, sizes, ... you can just add a dimension for each.

**Reply by isaacolsen94** [reply]
Thank you both! This was very helpful! Sorry for all the questions, is there a good spot that lists all (or a good majority) of the plugins available? I found the [Third-Part Intergrations](https://docs.inventree.org/en/stable/extend/integrate/) page, but that bulk plugin isn't mentioned? So is there like a bunch of plugin just only those few are mentioned?

**Reply by isaacolsen94** [reply]
So I've been trying for the last few hours to install the plugins, but it keeps failing. I believe for some reason the template directory that it's looking for the plugin.html file is mapped wrong. But I just did a brand new clean install and it is still having the same issue. This is the error I get when I try the RapidDocs plugin:

```
Traceback (most recent call last):

File "/root/.local/lib/python3.11/site-packages/django/core/handlers/base.py", line 220, in _get_response
response = response.render()
^^^^^^^^^^^^^^^^^

File "/root/.local/lib/python3.11/site-packages/django/template/response.py", line 114, in render
self.content = self.rendered_content
^^^^^^^^^^^^^^^^^^^^^

File "/root/.local/lib/python3.11/site-packages/rest_framework/response.py", line 70, in rendered_content
ret = renderer.render(self.data, accepted_media_type, context)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/root/.local/lib/python3.11/site-packages/rest_framework/renderers.py", line 159, in render
template = self.resolve_template(template_names)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/root/.local/lib/python3.11/site-packages/rest_framework/renderers.py", line 169, in resolve_template
return loader.select_template(template_names)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/root/.local/lib/python3.11/site-packages/django/template/loader.py", line 47, in select_template
raise TemplateDoesNotExist(", ".join(template_name_list), chain=chain)

django.template.exceptions.TemplateDoesNotExist: rapidoc.html
```

I can provide any info needed? But any help would be greatly appreciated!

![image](https://github.com/inventree/InvenTree/assets/52682311/d3d58d1f-ea19-428d-929b-d62c535dc0ca)

**Reply by wolflu05** [reply]
This is the official plugin list where anyone can submit their plugin. https://inventree.org/plugins

And regarding the rapidoc plugin, I'm not sure, maybe @matmair knows something.

**Reply by matmair** [reply]
Sounds like an issue with the rapidoc plugin - please [raise an issue in it's repo](https://github.com/invenhost/inventree-rapidoc) and I will try to diagnose it

**Reply by isaacolsen94** [reply]
Currently this is happening with all three of the plug-ins I installed in the GUI. The other two don't explicitly have the same error output. But when I go to the "plugin/plugin-name/" url it's just routes my bank to the home page.

**Reply by wolflu05** [reply]
@matmair i guess the issue is that the static and template folders are not in the manifest.in so I guess they are missing in the pip package.

And @isaacolsen94 please describe more what you did and what you think is wrong. And If you navigate to the categories and select one category, do you have the bulk create parts panel on the left?

**Reply by isaacolsen94** [reply]
I initially went in and added the bulk, rapid doc, and kintree plug-ins. I also enabled the url and app settings then restarted the docker stack. Then after enabling I went into each plug-in setting page and clicked the open in new tab option for each plugin options and that's when nothing happens or I get errors. I'll include screenshots below, but I think that for some reason it can't find the template/plugin.html file installed. It may be that there is other stuff wrong and it's just silently failing in the background. I don't currently have any categories setup yet, I've been trying to get this working so I could bulk add the category/parts. I'll try real fast though

**Comment by isaacolsen94** [comment]
https://github.com/inventree/InvenTree/assets/52682311/16925366-dfdf-4fd1-becb-d1c27ab3aa07


https://github.com/inventree/InvenTree/assets/52682311/b93152f4-f873-41f3-99ac-b3de1aef00bd


https://github.com/inventree/InvenTree/assets/52682311/e97c76b6-2dfc-4eb9-b5e7-627b1273e177

**Reply by isaacolsen94** [reply]
Bulk category/part do show up. So maybe it's just the other two or I'm my expected behavior is different from what the plugin is actually doing

**Reply by wolflu05** [reply]
You clicked on a base url for the plugins url endpoints. There is nothing registered. And secondly these three urls you see are api endpoints, you cannot click on them and expect that something shows up.

As I said before, goto a category (NOT the root category) and there should be a bulk create parts and bulk create category panel (same for stock with locations). 

And generally, you should read the readme of those plugins if you have no idea of how the plugins work.

(And note that the plugin you named kintree is not ki-n-tree, its a plugin which provides an InvenTree interface for the new kicad http libraries.)

**Reply by isaacolsen94** [reply]
Welp I'm an idiot. So I guess the issue is just something with rapidocs and the others are working as intended. Or is that error expected if I click from the plugin page?

**Reply by wolflu05** [reply]
@isaacolsen94 I think so that the error only exists on your setup is with the rapid doc plugin.

> Or is that error expected if I click from the plugin page?

You mean clicking on the baseurl for a plugin, or the big "Open in new tab" buttons? Yes, they are a bit misleading, at least for most plugins, that use the urls to define api routes. You cannot open these with the browser. So it is expected to "not work" show something useful.

If you want to use the bulk plugin to bulk create parts:
- goto the parts page
- create a category
- goto the categories page and there should be the panels "Bulk create part" and "Bulk create category"
- select the "bulk create parts" panel and click "New untitled schema"
- enter `*NUMERIC` in the dimensions field and `10` in the count field below. Then enter `Part {{dim.1}}` in name
- click "Preview", a list of generated parts with the name `Part 1` to `Part 10` should be shown
- and read more the various options on the readme: https://github.com/wolflu05/inventree-bulk-plugin?tab=readme-ov-file#-usage
